[ECO-5105] docs: updated `README.md` (#65)
ably-labs/ably-chat-kotlin: README.md

# Ably Chat SDK for Android

<p style="text-align: left">
    <img src="https://img.shields.io/badge/development_status-Private_Beta-ab7df8" alt="Development status"   />

>     <img src="https://badgen.net/github/license/3scale/saas-operator" alt="License" />

Reviewer: Can we also add one more badge for the current version, like
`ably-chat-kotlin:0.0.1-beta`

Author: done

</p>

Ably Chat is a set of purpose-built APIs for a host of chat features enabling you to create 1:1, 1:Many, Many:1 and Many:Many chat rooms for
any scale. It is designed to meet a wide range of chat use cases, such as livestreams, in-game communication, customer support, or social
interactions in SaaS products. Built on [Ably's](https://ably.com/) core service, it abstracts complex details to enable efficient chat
architectures.

> [!IMPORTANT]
> This SDK is currently under development. If you are interested in being an early adopter and providing feedback then you
> can [sign up to the private beta](https://forms.gle/vB2kXhCXrTQpzHLu5) and are welcome
> to [provide us with feedback](https://forms.gle/mBw9M53NYuCBLFpMA). Coming soon: chat moderation, editing and deleting messages.

Get started using the [📚 documentation](https://ably.com/docs/products/chat).

> ![Ably Chat Header](/images/ably-chat-github-header.png)

Author: oh, nice catch, forgot it

## Supported Platforms

> This SDK works on Android 7.0+ (API level 24+) and Java 8+.

Reviewer: While we support `Android 7.0+ (API level 24+) and Java 8+`. 
We don't have working CI environment ( to back our claim) for java 8 or tests running on emulator API level 24+
Maybe we should add support/ create separate issue for the same.

Author: I think we can address this later

## Supported chat features

This project is under development so we will be incrementally adding new features. At this stage, you'll find APIs for the following chat
features:

- Chat rooms for 1:1, 1:many, many:1 and many:many participation.
- Sending and receiving chat messages.
- Online status aka presence of chat participants.
- Chat room occupancy, i.e total number of connections and presence members.
- Typing indicators
- Room-level reactions (ephemeral at this stage)

If there are other features you'd like us to prioritize, please [let us know](https://forms.gle/mBw9M53NYuCBLFpMA).

## Usage

You will need the following prerequisites:

- An Ably account
    - You can [sign up](https://ably.com/signup) to the generous free tier.
- An Ably API key
    - Use the default or create a new API key in an app within
      your [Ably account dashboard](https://ably.com/dashboard).
    - Make sure your API key has the
      following [capabilities](https://ably.com/docs/auth/capabilities): `publish`, `subscribe`, `presence`, `history` and
      `channel-metadata`.

> [//]: # (## Installation)

Reviewer 2: Is this intentionally commented?

Author: Yes, I think we should release it in the Maven first

[//]: # ()
[//]: # (Certain functionalities are powered by the [ably-android]&#40;https://github.com/ably/ably-java&#41; library. )

[//]: # (The `ably-android` library is included as an api dependency within the Chat SDK, so there is no need to manually add it to your project.)

## Versioning

The Ably client library follows [Semantic Versioning](http://semver.org/). See https://github.com/ably/ably-chat-kotlin/tags for a list of
tagged releases.

## Instantiation and authentication

To instantiate the Chat SDK, create an [Ably client](https://ably.com/docs/getting-started/setup) and pass it into the
Chat constructor:

```kotlin
import com.ably.chat.ChatClient
import io.ably.lib.realtime.AblyRealtime
import io.ably.lib.types.ClientOptions

val realtimeClient = AblyRealtime(
    ClientOptions().apply {
        key = "<api-key>"
        clientId = "<client-id>"
    },
)

val chatClient = ChatClient(realtimeClient)
```

You can use [basic authentication](https://ably.com/docs/auth/basic) i.e. the API Key directly for testing purposes,
however it is strongly recommended that you use [token authentication](https://ably.com/docs/auth/token) in production
environments.

> To use Chat you must also set a [`clientId`](https://ably.com/docs/auth/identified-clients) so that clients are
> identifiable. If you are prototyping, you can use `java.util.UUID` to generate an ID.

Reviewer: ```suggestion
To use Chat you must also set a [`clientId`](https://ably.com/docs/auth/identified-clients) so that clients are
identifiable. If you are prototyping, you can use `java.util.UUID` to generate an ID.
P.S. In most cases, you can set `clientId` to `userId`, provided that `userId` is unique for your application.
```

Author: Don't want to overcomplicate here, we never mentioned `userId` before, I think it can confuse the SDK users. And not sure we should mention something about `userId` being unique.

Reviewer: Although, we never mention it before, that's the whole point of having `clientId`. 
From user perspective, they shouldn't randomly set clientId as UUID, rather it should be mapped to corresponding user or use `userId` instead if it's unique

Reviewer: You can ignore this thread for now, I will start a different thread on slack.
since this has overall chat docs usage implications across SDKs
Cc @AndyTWF

> ## Connections

Reviewer: We should either remove or mark this section as in progress, since we don't have any impl. for the `Connection`.
Pending issue for the same -> https://github.com/ably-labs/ably-chat-kotlin/issues/63

Reviewer: Also, external shared realtime instance already provide a way to listen to `connection` events
https://github.com/ably/ably-java?tab=readme-ov-file#connection

The Chat SDK uses a single connection to Ably, which is exposed via the `ChatClient#connection` property. You can use this
property to observe the connection state and take action accordingly.

### Current connection status

You can view the current connection status at any time:

```kotlin
val connectionStatus = chat.connection.status
val connectionError = chat.connection.error
```

### Subscribing to connection status changes

You can subscribe to connection status changes by registering a listener, like so:

```kotlin
val subscription = chat.connection.onStatusChange { statusChange: ConnectionStatusChange ->
    println(statusChange.toString())
}
```

> To stop listening to changes, `unsubscribe` method on returned subscription instance:

Reviewer 2: ```suggestion
To stop listening to changes, call the `unsubscribe` method on the returned subscription instance:
```

```kotlin
subscription.unsubscribe()
```

## Chat rooms

### Creating or retrieving a chat room

You can create or retrieve a chat room with name `"basketball-stream"` this way:

```kotlin

> val room = chat.rooms.get("basketball-stream", RoomOptions(reactions = RoomReactionsOptions()))

Reviewer 2: Would it be worth an example with, say, a custom typing timeout, so the customisability is shown?

Author: done

```

> The second argument to `rooms.get` is a `RoomOptions` argument, which tells the Chat SDK what features you would like your room to use and
> how they should be configured.

Reviewer: ```suggestion
The second argument for `rooms.get` is a `RoomOptions` parameter, which specifies the features you want your room to utilize and their configuration. 
P.S. You can also use `RoomOptions.default` to enable all room features with the default configuration.
```

### Attaching to a room

To start receiving events on a room, it must first be attached. This can be done using the `attach` method.

```kotlin
// Add a listener so it's ready at attach time (see below for more information on listeners)
room.messages.subscribe { message: Message ->
    println(message.toString())
}

room.attach()
```

### Detaching from a room

To stop receiving events on a room, it must be detached, which can be achieved by using the `detach` method.

```kotlin
room.detach()
```

Note: This does not remove any event listeners you have registered and they will begin to receive events again in the
event that the room is re-attached.

### Releasing a room

Depending on your application, you may have multiple rooms that come and go over time (e.g. if you are running 1:1 support chat). When you
are completely finished with a room, you may `release` it which allows the underlying resources to be collected.

```kotlin
rooms.release("basketball-stream")
```

Once `release` is called, the room will become unusable and you will need to get a new instance using `rooms.get` should you wish to
re-start the room.

> Note that releasing a room may be optional for many applications.

Reviewer 2: As we're going to mention not requiring release, perhaps worth mentioning that the server will tidy up connections etc after a period of time?

Author: done

### Monitoring room status

Monitoring the status of the room is key to a number of common chat features. For example, you might want to display a warning when the room
has become detached.

### Current status of a room

To get the current status, you can use the `status` property:

```kotlin
val roomStatus = room.status
val roomError = room.error
```

### Listening to room status updates

You can also subscribe to changes in the room status and be notified whenever they happen by registering a listener:

```kotlin

> val subscription = room.onStatusChange { statusChange: RoomStatusChange ->

Reviewer 2: Not for this PR to fix, but OOI did we consider using `StateFlows` for this purpose - I'd imagine they fit quite well with the status change paradigm?

Reviewer: With current Api design, we have problem with unsubscribing to state flows . Since,  it's not deterministic in nature : ( 
Maybe we can expose flow as an extension method in the future alongside current public API

Reviewer 2: Ah, fair enough - that makes sense, so it's like you call `unsubscribe` but there's no guarantees that that will happen immediately, or that you won't receive any other events in the time?

But that's reason enough to not do this now, I was just curious - thanks for the explanation. We can definitely consider in the future and I would agree that it's an extension to the current API, if we choose to do this :)

```

### Unsubscribing from incoming messages

When you're done with the listener, call `unsubscribe` to remove that listeners subscription and prevent it from receiving
any more events.

```kotlin
val subscription = room.messages.subscribe { message: Message ->
    println(message.toString())
}
// Time passes...
subscription.unsubscribe()
```

### Retrieving message history

The messages object also exposes the `get` method which can be used to request historical messages in the chat room according
to the given criteria. It returns a paginated response that can be used to request more messages.

```kotlin
var historicalMessages = room.messages.get(QueryParams(orderBy = NewestFirst, limit = 50))
println(historicalMessages.items.toString())

while (historicalMessages.hasNext()) {
    historicalMessages = historicalMessages.next()
    println(historicalMessages.items.toString())
}

println("End of messages")
```

### Retrieving message history for a subscribed listener

In addition to being able to unsubscribe from messages, the return value from `messages.subscribe` also includes the `getPreviousMessages`
method. It can be used to request

> historical messages in the chat room that were sent up to the point a that particular listener was subscribed. It returns a

Reviewer 2: ```suggestion
historical messages in the chat room that were sent up to the point that a particular listener was subscribed. It returns a
```

action. While leaving presence, you can provide optional data that will be associated with the presence message.

```kotlin
room.presence.leave(
    JsonObject().apply {
        addProperty("status", "Be back later!")
    },
)
```

### Subscribing to presence updates

You can provide a single listener, if so, the listener will be subscribed to receive all presence event types.

```kotlin
val subscription = room.presence.subscribe { event: PresenceEvent ->
    when (event.action) {
        Action.enter -> println("${event.clientId} entered with data: ${event.data}")
        Action.leave -> println("${event.clientId} left")
        Action.update -> println("${event.clientId} updated with data: ${event.data}")
    }
}
```

### Unsubscribing from presence updates

To unsubscribe a specific listener from presence events, you can call the `unsubscribe` method provided in the subscription object returned
by the `subscribe` call.

```kotlin
val subscription = room.presence.subscribe { event: PresenceEvent ->
    // Handle events
}

// Unsubscribe
subscription.unsubscribe()
```

## Typing indicators

> Typing events allow you to inform others that a client is typing and also subscribe to others' typing status.

Reviewer 2: We should include that you need to be attached to the room to type

Author: done
